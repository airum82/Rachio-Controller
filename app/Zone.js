import React, { Component } from 'react';
import { apiKey } from '../APIkey';

class Zone extends Component {
  constructor() {
    super();
    this.state = {
      showForm: false,
      duration: 0,
      selected: false,
      error: ''
    };
    this.toggleForm = this.toggleForm.bind(this);
    this.startZone = this.startZone.bind(this);
    this.handleInput = this.handleInput.bind(this);
    this.handleSelect = this.handleSelect.bind(this);
  }

  toggleForm() {
    this.setState({ showForm: !this.state.showForm });
  }

  handleInput(event) {
    this.setState({
      [event.target.name]: event.target.value
    })
  }

  handleSelect() {
    this.props.selectZone(this.props.id)
    this.setState({ selected: !this.state.selected })
  }

  startZone(event) {
    event.preventDefault();
    const body = {
      id: this.props.id,
      duration: parseInt(this.state.duration)
    };
    return fetch('https://api.rach.io/1/public/zone/start', {
      method: 'PUT',
      headers: {
        'Authorization': `Bearer ${apiKey}`,
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(body)
    })
    .then(response => response.text())
    .then(result => console.log(result))
    .then(() => this.toggleForm())
    .catch(error => this.setState({ error: error.message }))
  }

  render() {
    return (
      <div className="zone" id={this.props.id}>
        <h3 className="zone-title">{this.props.name}</h3>
        <img 
          className={this.state.selected ? 'selected' : ''}
          onClick={this.handleSelect}
          src={this.props.image} 
          alt="picture of this particular zone" 
        />
        <div className="zone-buttons">
          <button 
            onClick={this.toggleForm}
            className="toggle-form"
          >
            {this.state.showForm ? 'cancel' : 'start zone'}
          </button>
          <button 
            onClick={this.handleSelect}
            className="select-zone"
          >
            {this.state.selected ? 'unselect' : 'select zone'}
          </button>
        </div>
       <form 
        className={this.state.showForm ? 'zone-form' : 'hidden-form'}
        onSubmit={this.startZone}
       >
        <h4>duration</h4>
        <input name="duration" type="text" onChange={this.handleInput}/>
        <button>start</button>
       </form>
      </div>
    )
  }
}

export default Zone;