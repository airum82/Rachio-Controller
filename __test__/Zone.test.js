import React from 'react';
import Zone from '../app/Zone';
import { shallow } from 'enzyme';

describe('Zone', () => {
  let wrapper;
  const mockProps = {
    name: 'zone-4000',
    id: '45-7c',
    selectZone: jest.fn(),
    image: '',
    enabled: true,
    addSortOrder: jest.fn()
  }
  beforeEach(() => {
    wrapper = shallow(<Zone
      name={mockProps.name}
      id={mockProps.id}
      selectZone={mockProps.selectZone}
      image={mockProps.image}
      enabled={mockProps.enabled}
      addSortOrder={mockProps.addSortOrder}
    />)
  })
  it('should match snapshot upon initial render', () => {
    expect(wrapper).toMatchSnapshot();
  })
  it('toggleForm should set state with opposite boolean value if zone is enabled', () => {
    const initialValue = wrapper.state().showForm;
    wrapper.instance().toggleForm();
    expect(!wrapper.state().showForm).toEqual(initialValue);
  })
  it('toggle form should call handle disabled if zone is disabled', () => {
    wrapper = shallow(<Zone
      name="I'm not working"
      id="not-me"
      selectZone={jest.fn()}
      image={''}
      enabled={false}
       />)
    const spy = jest.spyOn(wrapper.instance(), 'handleDisabled');
    wrapper.instance().toggleForm();
    expect(spy).toHaveBeenCalled();
  })
  it('handleInput should set state with value of input', () => {
    wrapper.instance().toggleForm();
    const mockEvent = {
      target: {
        value: 'hi',
        name: 'bob'
      }
    }
    const spy = jest.spyOn(wrapper.instance(), 'setState');
    wrapper.instance().handleInput(mockEvent);
    expect(spy).toHaveBeenCalledWith({ bob: 'hi' });  
  })
  it('handle select should set state with opposite boolean value of select', () => {
    const initialValue = wrapper.state().selected;
    wrapper.instance().handleSelect();
    expect(!wrapper.state().selected).toEqual(initialValue)
  })
  it('handle select should call handleDisabled if zone is disabled', () => {
    wrapper = shallow(<Zone
      name="I'm not working"
      id="not-me"
      selectZone={jest.fn()}
      image={''}
      enabled={false}
    />)
    const spy = jest.spyOn(wrapper.instance(), 'handleDisabled');
    wrapper.instance().handleSelect();
    expect(spy).toHaveBeenCalled();
  })
  it('grabSortOrder should call addSortOrder with correct params', () => {
    const mockEvent = {
      target: {
        value: '5'
      }
    }
    wrapper.instance().grabSortOrder(mockEvent);
    expect(mockProps.addSortOrder).toHaveBeenCalledWith(mockProps.id, mockEvent.target.value);
  })
  it('startZone should call fetch', () => {
    const mockEvent = {
      preventDefault: jest.fn()
    }
    window.fetch = jest.fn().mockImplementation(() => Promise.resolve({
      text: () => Promise.resolve({})
    }))
    wrapper.instance().startZone(mockEvent);
    expect(window.fetch).toHaveBeenCalled();
  })
  it('startZone should call toggleForm', () => {
    window.fetch = jest.fn().mockImplementation(() => Promise.resolve({
      text: () => Promise.resolve({})
    }));
    const mockEvent = {
      preventDefault: jest.fn()
    }
    const spy = jest.spyOn(wrapper.instance(), 'toggleForm');
    return wrapper.instance().startZone(mockEvent)
      .then(() => {
        expect(spy).toHaveBeenCalled();
      })
  })
  it('startZone should set state with message of fetch fails', () => {
    window.fetch = jest.fn().mockImplementation(() => Promise.reject({
      message: 'Could not start zone'
    }))
    const mockEvent = {
      preventDefault: jest.fn()
    }
    const spy = jest.spyOn(wrapper.instance(), 'setState');
    return wrapper.instance().startZone(mockEvent)
      .then(() => {
        expect(spy).toHaveBeenCalledWith({ error: 'Could not start zone' });
      })
  })
  it('toggle-form button should call toggleForm', () => {
    const spy = jest.spyOn(wrapper.instance(), 'toggleForm');
    wrapper.find('.toggle-form').simulate('click');
    wrapper.find('.toggle-form').simulate('click');
    expect(spy).toHaveBeenCalled();
  })
  it('image should call handleSelect when clicked', () => {
    const spy = jest.spyOn(wrapper.instance(), 'handleSelect');
    wrapper.find('img').simulate('click');
    wrapper.find('img').simulate('click');
    expect(spy).toHaveBeenCalled();
  })
  it('select button should call handleSelect when clicked', () => {
    const spy = jest.spyOn(wrapper.instance(), 'handleSelect');
    wrapper.find('.select-zone').simulate('click');
    wrapper.find('.select-zone').simulate('click');
    expect(spy).toHaveBeenCalled();
  })
  it('input field should call handleInput on change', () => {
    const spy = jest.spyOn(wrapper.instance(), 'handleInput');
    wrapper.find('.duration-input').simulate('change', {
      target: {
        name:'jenny',
        value: 'hey'
      }
    })
    wrapper.find('.duration-input').simulate('change', {
      target: {
        name: 'jenny',
        value: 'hey'
      }
    })
    expect(spy).toHaveBeenCalled();
  })
  it('sort-order input should call grabSortOrder on change', () => {
    const spy = jest.spyOn(wrapper.instance(), 'grabSortOrder');
    wrapper.instance().setState({ selected: true });
    const mockEvent = {
      target: {
        value: '5'
      }
    }
    wrapper.find('.sort-order').simulate('change', mockEvent)
    wrapper.find('.sort-order').simulate('change', mockEvent)
    expect(spy).toHaveBeenCalledWith(mockEvent);
  })
})