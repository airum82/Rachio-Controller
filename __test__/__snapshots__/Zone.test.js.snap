// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Zone should match snapshot upon initial render 1`] = `
<div
  className="zone"
  id="45-7c"
>
  <h3
    className="zone-title"
  >
    zone-4000: enabled
  </h3>
  <img
    alt="picture of this particular zone"
    className=""
    onClick={[Function]}
    src=""
  />
  <div
    className="zone-buttons"
  >
    <button
      className="toggle-form"
      onClick={[Function]}
    >
      start zone
    </button>
    <button
      className="select-zone"
      onClick={[Function]}
    >
      select zone
    </button>
  </div>
  <form
    className="hidden-form"
    onSubmit={[Function]}
  >
    <h4>
      duration(in seconds):
    </h4>
    <input
      className="duration-input"
      name="duration"
      onChange={[Function]}
      type="text"
    />
    <button>
      start
    </button>
  </form>
</div>
`;
